devtools/enforcer-rules/src/main/java/io/quarkus/enforcer/RequiresMinimalDeploymentDependency.java

```diff
@@ -0,0 +1,146 @@
+
+        Map<String, Dependency> directQuarkusDepsByGAV = project.getDependencies().stream()
+                .filter(d -> d.getGroupId().equals(GROUP_ID_FILTER))
+                .collect(Collectors.toMap(d -> d.getGroupId() + ":" + d.getArtifactId() + ":" + d.getVersion(), d -> d));
+
+        Map<String, Artifact> nonDeploymentQuarkusArtifactsByGAV = project.getArtifacts().stream()
+                .filter(artifact -> artifact.getGroupId().equals(GROUP_ID_FILTER))
+                .filter(artifact -> !artifact.getArtifactId().endsWith("-deployment"))
+                .collect(Collectors.toMap(a -> a.getGroupId() + ":" + a.getArtifactId() + ":" + a.getVersion(), a -> a));
+
+        String missingDeploymentDeps = nonDeploymentQuarkusArtifactsByGAV.entrySet().parallelStream()
+                .filter(entry -> directQuarkusDepsByGAV.containsKey(entry.getKey())) // only direct deps
+                .map(entry -> DEPLOYMENT_GAV_CACHE.computeIfAbsent(entry.getKey(), k -> parseDeploymentGAV(entry.getValue())))
+                .filter(optDeploymentGAV -> optDeploymentGAV
+                        .map(deploymentGAV -> !isMinimalDeploymentDepPresent(deploymentGAV, directQuarkusDepsByGAV))
+                        .orElse(false))
+                .map(Optional::get)
+                .sorted()
+                .collect(Collectors.joining("\n    "));
+
+        if (!missingDeploymentDeps.isEmpty()) {
+            throw new EnforcerRuleException("The following *-deployment dependencies are missing/configured incorrectly:\n"
+                    + "    " + missingDeploymentDeps
+                    + "\n\nEach of these dependencies has to be configured as follows to get a consistent build order:\n"
+                    + "    <dependency>\n"
+                    + "        <groupId>GROUP-IP</groupId>\n"
+                    + "        <artifactId>ARTIFACT-ID</artifactId>\n"
+                    + "        <version>${project.version}</version>\n"
+                    + "        <type>pom</type>\n"
+                    + "        <scope>test</scope>\n"
+                    + "        <exclusions>\n"
+                    + "            <exclusion>\n"
+                    + "                <groupId>*</groupId>\n"
+                    + "                <artifactId>*</artifactId>\n"
+                    + "            </exclusion>\n"
+                    + "        </exclusions>\n"
+                    + "    </dependency>");
+        }
+    }
+
+    private Optional<String> parseDeploymentGAV(Artifact artifact) {
+        if (!artifact.isResolved()) {
+            // To avoid this "soft exit", explicit resolving would be necessary but that is pretty elaborate in an enforcer rule.
+            // If the build goal is "late" enough, artifacts for the respective scope *will* be resolved automatically.
+            logger.warn("Artifact is not resolved, consider using a later build goal like 'package': " + artifact);
```

Comment:
Example:
```
$ mvn clean validate -f integration-tests/cache/
[INFO] Scanning for projects...
[INFO]
[INFO] -------------< io.quarkus:quarkus-integration-test-cache >--------------
[INFO] Building Quarkus - Integration Tests - Cache 999-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-clean-plugin:3.1.0:clean (default-clean) @ quarkus-integration-test-cache ---
[INFO]
[INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (enforce) @ quarkus-integration-test-cache ---
[INFO]
[INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-java-version) @ quarkus-integration-test-cache ---
[INFO]
[INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-maven-version) @ quarkus-integration-test-cache ---
[INFO]
[INFO] --- maven-enforcer-plugin:3.0.0-M3:enforce (enforce-deployment-deps) @ quarkus-integration-test-cache ---
[WARNING] Artifact is not resolved, consider using a later build goal like 'package': io.quarkus:quarkus-junit5:jar:999-SNAPSHOT:test
[WARNING] Artifact is not resolved, consider using a later build goal like 'package': io.quarkus:quarkus-resteasy-jsonb:jar:999-SNAPSHOT:compile
[WARNING] Artifact is not resolved, consider using a later build goal like 'package': io.quarkus:quarkus-hibernate-orm-panache:jar:999-SNAPSHOT:compile
[WARNING] Artifact is not resolved, consider using a later build goal like 'package': io.quarkus:quarkus-cache:jar:999-SNAPSHOT:compile
[WARNING] Artifact is not resolved, consider using a later build goal like 'package': io.quarkus:quarkus-smallrye-context-propagation:jar:999-SNAPSHOT:compile
[WARNING] Artifact is not resolved, consider using a later build goal like 'package': io.quarkus:quarkus-jdbc-h2:jar:999-SNAPSHOT:compile
[WARNING] Artifact is not resolved, consider using a later build goal like 'package': io.quarkus:quarkus-test-h2:jar:999-SNAPSHOT:test
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

The warnings vanish when `test` is used instead of `validate` and the build will fail (because some deployment deps are missing).

This is not ideal, but I'd say it is sufficient for what the rule is supposed to do, which is to enforce the build order primarily in CI.
CI will always use a "late" phase/goal (at least test).

Reply:
@famod , do I understand it correctly `test`  vs `validate` makes difference only if you build the whole project and not when you only select `-f integration-tests/cache/`?

Reply:
It also makes a difference when builing the entire project.
`mvn validate` called from root would then log many warnings, but failing the build just because artifacts have not yet been resolved would be worse, IMO.

To me, this is an edge case, because usually everyone (including CI) is at least using `compile`, `test`, `package` or even `install`.

I could also squash this into a single warning making it less noisy. WDYT?

Reply:
I squashed it to a single message for less noise, e.g.:
```
[INFO] --- maven-enforcer-plugin:3.0.0-M2:enforce (enforce-deployment-deps) @ quarkus-integration-test-cache ---
[WARNING] Skipping rule RequiresMinimalDeploymentDependency: Artifacts are not resolved, consider using a later build goal like 'package'.
```